web-client/public/js/upload.js

```diff
@@ -121,96 +245,71 @@ export const upload = function () {
-                                    finalExportSheets.expression[sheet] = null;
-                                }
-                            });
-                            grnState.workbook.exportSheets = finalExportSheets;
-                            if (!$(this).parent().hasClass("disabled")) {
-                                var workbookToExport = flattenWorkbook(uploadState.currentWorkbook, sheetType);
-                                var workbookFilename = filenameWithExtension(sheetType !== uploadState.currentWorkbook.sheetType ?
-                                    sheetType : "", extension);
-                                workbookToExport.filename = workbookFilename;
-                                var exportForm = $("<form></form>").attr({
-                                    method: "POST",
-                                    action: $(".service-root").val() + "/" + route
-                                }).append($("<input></input>").attr({
-                                    type: "hidden",
-                                    name: "filename",
-                                    value: workbookFilename
-                                })).append($("<input></input>").attr({
-                                    type: "hidden",
-                                    name: "workbook",
-                                    value: JSON.stringify(workbookToExport)
-                                }));
-                                $("body").append(exportForm);
-                                exportForm.submit();
-                                exportForm.remove();
-                            }
-                            $("#exportExcelModal").modal("hide");
+                    queryExpressionDatabase({
+                        type:twoColumnSheetType[sheet],
+                        genes:grnState.workbook.genes.map(x => {
+                            return x.name;
+                        }).join(",")
+                    }).then(function (response) {
+                        result.data = response;
+                        finalExportSheets.two_column_sheets[sheet] = result;
+                        if (!Object.values(finalExportSheets.two_column_sheets).includes(null)) {
+                            // if we got all of the two column sheets, then proceed with export
+                            handleExpressionDataAndExport(route, extension, sheetType, source, finalExportSheets);
                         }
-                    }
-                }).catch(function (error) {
-                    console.log(error.stack);
-                    console.log(error.name);
-                    console.log(error.message);
-                });
+
+                    }).catch(function (error) {
+                        console.log(error.stack);
+                        console.log(error.name);
+                        console.log(error.message);
+                    });
```

Comment:
Second appearance of this function—can be defined as some standard `errorHandler` so that you would just need to say `catch(errorHandler)` here

Reply:
I created a function called `expressionExportErrorHandler` that is used in the functions `handleExpressionDataAndExport` and `handleExportExcelButtonExport`. It is called like this: `.catch((error) => expressionExportErrorHandler(error))`